### PR TITLE
Add instrumentation patch metrics

### PR DESCRIPTION
**Pull request by delner**
As a further extension of #867, this pull request refactors how integration patching works, and adds a metric for whenever integration patching succeeds/fails.

New metrics include:

- **datadog.tracer.instrumentation_patched**: Whenever a `Datadog::Contrib::Patcher#patch` succeeds.
  - *patcher*: Patcher that was applied.
  - *target_version*: Version of software that was patched (typically a gem version.)
- **datadog.tracer.error.instrumentation_patch**: Whenever a `Datadog::Contrib::Patcher#patch` raises an error.
  - *patcher*: Patcher that was attempted.
  - *target_version*: Version of software that patch was attempted against (typically a gem version.)
  - *error*: Class name of error raised.

Screenshot from 2019-12-02 18-26-22

### CODE REVIEW

File: lib/ddtrace/contrib/action_pack/action_controller/patcher.rb [hunk 1]
```diff
@@ -11,24 +11,8 @@ module Patcher
 
           module_function
 
-          def patched?
-            done?(:action_controller)
-          end
-
           def patch
-            do_once(:action_controller) do
```
Comment (marcotc) [inline lib/ddtrace/contrib/action_pack/action_controller/patcher.rb]: Are these safe-guards not needed anymore?

Comment (delner) [reply lib/ddtrace/contrib/action_pack/action_controller/patcher.rb]: No they are, they're rolled into `Datadog::Contrib::Patcher` so they're automatic, so to say.

**Comment by delner** [comment]
This is now ready for review.

I think this is fine as is, but it might be worthwhile also adding a "gem version" tag to the metric of the library that failed to patch. e.g. If our Rails integration failed to patch Rails 6.1, it would add a `target_version:6.1.0` tag to the error metric.

File: lib/ddtrace/contrib/patcher.rb [hunk 2]
```diff
@@ -6,21 +6,38 @@ module Contrib
     module Patcher
       def self.included(base)
         base.send(:include, Datadog::Patcher)
-        base.send(:extend, InstanceMethods)
-        base.send(:include, InstanceMethods)
+
+        base.singleton_class.send(:prepend, CommonMethods)
+        base.send(:prepend, CommonMethods) if base.class == Class
```
Comment (marcotc) [inline lib/ddtrace/contrib/patcher.rb]: Any particular reason for handling the case where the Patcher is a class, given all our current Patchers are modules?
Would it be for future proofing?

Comment (delner) [reply lib/ddtrace/contrib/patcher.rb]: Yeah, technically our integration API doesn't require all patchers to be modules, just that they include `Datadog::Contrib::Patcher` and behave like one. This means users who have created custom integrations might have implemented the patchers as classes: it's more of a "avoiding possible breaking change" thing than future proofing.

**Comment by delner** [comment]
Added `target_version` tag and an `instrumentation_patched` metric which is emitted whenever an integration's patch succeeds.

**Pull request merged**